parsedmarc: fix and switch to OpenSearch

talyz opened the pull request:
## Description of changes

### Problems
**parsedmarc** has been broken for a while now, due to incompatible updates of the Elasticsearch python packages. This probably wasn't very obvious to anyone as the tests require `allowUnfree`, since Elasticsearch is no longer open source.

### Solutions
- Fix the broken python dependencies by pinning compatible ones
- Switch from Elasticsearch to OpenSearch

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

talyz committed:
parsedmarc: Fix runtime errors

parsedmarc has been broken for a while, since the elasticsearch python
packages were upgraded past its supported versions. To fix this, pin
the versions locally.

talyz committed:
nixos/parsedmarc: Switch from Elasticsearch to OpenSearch

Switch to OpenSearch, an open source and compatible fork of
Elasticsearch, as the default locally deployed search engine. This
means it's once again possible to run parsedmarc without unfree
packages and the tests should therefore run in CI.

dotlambda commented on pkgs/development/python-modules/parsedmarc/default.nix:
```diff
@@ -1,3 +1,24 @@
+{ pythonPackages }:
+let
+  packages = pythonPackages.overrideScope (final: prev: {
```
We can't override within `pythonPackages`.
Should we set `meta.broken = true` if the elasticsearch or elasticsearch-dsl version is too high? I assume most of the Python API still works.
We can override the package in `all-packages.nix` or in the NixOS module.

talyz replied:
Why is that? It works just fine and seems to be done in a few packages already.

Maybe? I don't use it any other way than as a server.

Overriding in `all-packages.nix` feels like a definite anti-pattern, since it means what you're seeing in the package file isn't what you expect. The sane alternative I can think of is to create new packages with the version pinned, but that seemed a bit unnecessary for just one package.

dotlambda replied:
> Why is that? It works just fine and seems to be done in a few packages already.

You cannot have two versions of a Python module in the same environment. That could happen if parsedmarc is used together with another module that uses the elasticsearch module.

dotlambda commented on nixos/modules/services/search/opensearch.nix:
```diff
@@ -37,7 +37,7 @@ in
           default = "127.0.0.1";
           description = lib.mdDoc ''
-            Which port this service should listen on.
+            Which address this service should listen on.
```
Can we put this in a separate commit?

talyz replied:
Sure, no problem!

talyz committed:
nixos/opensearch: Fix network.host option description

talyz committed:
nixos/parsedmarc: Fix apply for smtp.to

An incomplete implementation which left out null values was added in
https://github.com/NixOS/nixpkgs/pull/246992. Complete it by adding
the missing check.
